### ISSUE TITLE
find_nodes_in_area() volume calculation is off by about four million nodes, can crash or hang Minetest

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#11769]
##### Minetest version
Commit 80d12dbed

##### OS / Hardware
Operating system: unknown
CPU: unknown

##### Summary
Last night, some playful kids crashed Oysterity (running Mineclonia) in two different ways.
You can read all about it here: https://git.minetest.land/Mineclonia/Mineclonia/issues/165

Both crashes contained `find_nodes_in_area(): area volume exceeds allowed value of 4096000`.

@appgurueu suggested the volume calculation is wrong for coordinates exceeding s16 boundaries.

I can prove this with the following mod, the included `/areatest` chat command crashes Minetest.
It also prints out the coordinates that you have to give to `find_nodes_in_area()` to reproduce.

##### Steps to reproduce

1. Enter a minetest world with the `badarea` mod from the following zip file active:

data:application/zip;base64,UEsDBAoAAAAAABuNbFMAAAAAAAAAAAAAAAAIABwAYmFkYXJlYS9VVAkAAxaZjmFWno5hdXgLAAEE6AMAAAToAwAAUEsDBBQAAAAIAOSGbFMsW+VyaAAAAIUAAAAQABwAYmFkYXJlYS9tb2QuY29uZlVUCQADW46OYWSOjmF1eAsAAQToAwAABOgDAAAVjTEKAzEMBHu9QrgP3LlPkfxEZwtskO1D0kGeHwu2GmZ3Jw3GN15USZmgshXtt/c1N00pgbduOFZF/t3L2ELFcLGQlEcoXINQlYXJYi4f+XydOxm8uwT67tonLujxtnQTVuE2aE74A1BLAwQUAAAACAAbjWxTQWLKlQUBAABKAgAAEAAcAGJhZGFyZWEvaW5pdC5sdWFVVAkAAxaZjmEamY5hdXgLAAEE6AMAAAToAwAAlZBBboMwEEXXcIqRVyAZRFu16YazoAl2UkvxODKmMkS5e21kSojURRe29eb/8R+NViSdHFxt5VkNTtqu/0LXG62RRMHQSowy43DLTyP10EJ8nDJUEGpZ5idjwYfy2+t70/DwHD4OHF4aECbPojgFsUpqleQq6Yth/qs7yy6mxwtYFGocgutzq11NLNzAt57D1E4c5naG+2aIo2tFwfQte2dsPYxHZ7F3RWjl6c8y+vW6AyGP47mIpYyldsYXTFT7PU57nCOVTyOg30ZAIf6Rjv4xHf0uPeC0x+d0WpKIRh0G+E05KRIdGSGHTlEXe5dIgHVh/BHRr8hQWbYmSBL5cq3nXuY/UEsBAh4DCgAAAAAAG41sUwAAAAAAAAAAAAAAAAgAGAAAAAAAAAAQAO1BAAAAAGJhZGFyZWEvVVQFAAMWmY5hdXgLAAEE6AMAAAToAwAAUEsBAh4DFAAAAAgA5IZsUyxb5XJoAAAAhQAAABAAGAAAAAAAAQAAAKSBQgAAAGJhZGFyZWEvbW9kLmNvbmZVVAUAA1uOjmF1eAsAAQToAwAABOgDAABQSwECHgMUAAAACAAbjWxTQWLKlQUBAABKAgAAEAAYAAAAAAABAAAApIH0AAAAYmFkYXJlYS9pbml0Lmx1YVVUBQADFpmOYXV4CwABBOgDAAAE6AMAAFBLBQYAAAAAAwADAPoAAABDAgAAAAA=

2. Execute the `/areatest` command.

3. Get a crash similar to the following:

```
2021-11-12 17:41:04: [Server]: areamin	32492	-32508	32752
2021-11-12 17:41:04: [Server]: areamax	32508	-32492	32768
2021-11-12 17:41:04: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod 'badarea' in callback on_chat_message(): ...hare/src/games/minetest/bin/../mods/badarea/init.lua:22: find_nodes_in_area(): area volume exceeds allowed value of 4096000
2021-11-12 17:41:04: ERROR[Main]: stack traceback:
2021-11-12 17:41:04: ERROR[Main]: 	[C]: in function 'find_nodes_in_area'
2021-11-12 17:41:04: ERROR[Main]: 	...hare/src/games/minetest/bin/../mods/badarea/init.lua:22: in function 'func'
2021-11-12 17:41:04: ERROR[Main]: 	...hare/src/games/minetest/bin/../builtin/game/chat.lua:79: in function '?'
2021-11-12 17:41:04: ERROR[Main]: 	.../src/games/minetest/bin/../builtin/game/register.lua:425: in function <.../src/games/minetest/bin/../builtin/game/register.lua:409>
```

**Comment by erlehmann** [comment]
In response to the following message from #minetest-dev:
```
<sfan5> posting reproduction code as a base64 encoded ZIP file instead of plaintext should be criteria for immediate issue closure
```

I want to say that I wanted to attach the mod to reproduce it as a zip file but GitHub does not want it.

Putting the data URL in your browser address bar should enable you to “download” the zip file locally.

**Comment by erlehmann** [comment]
```
minetest.register_chatcommand("areatest", {
func = function(name)
for x = 32506, 32767, 10 do
	for y = -32507, -32768, -10 do
		for z = 32506, 32767, 10 do
			local radius = 8
			local pos = { x=x, y=y, z=z }
			local areamin = vector.subtract(pos, radius)
			minetest.debug(
				"areamin",
				areamin.x,
				areamin.y,
				areamin.z
			)
			local areamax = vector.add(pos, radius)
			minetest.debug(
				"areamax",
				areamax.x,
				areamax.y,
				areamax.z
			)
			local npos, nnum = minetest.find_nodes_in_area(
			   areamin,
			   areamax,
			   "air"
			)
		end
	end
end
end
})
```

**Comment by erlehmann** [comment]
It seems that if you manage to give exactly the s16 boundaries (-32768 or 32767) as an argument to `minetest.find_nodes_in_area()`, the server will hang at shutdown.

**Comment by erlehmann** [comment]
I am working on a fix.

**Comment by erlehmann** [comment]
Quoting myself from https://git.minetest.land/Mineclonia/Mineclonia/issues/165#issuecomment-28366 …

I may have a fix. Note that I advocate clamping the coordinates for the calculation here *only* because it is arguably correct to say that there are no nodes there … except if you want to read server RAM from Lua, maybe?

```
function clamp(value, min, max)
	assert( min < max )
	if value < min then
		return min
	end
	if value > max then
		return max
	end
	return value
end

assert( clamp(000, -100, 100) == 000 )
assert( clamp(999, -100, 100) == 100 )
assert( clamp(999, -999, 999) == 999 )
assert( clamp(998, 999, 1999) == 999 )
assert( clamp(999, 999, 1999) == 999 )

function clamp_s16(value)
	-- seems minetest hangs on -32768 and 32767
	return clamp(value, -32767, 32766)
end

assert( clamp_s16(000000) == 000000 )
assert( clamp_s16(000001) == 000001 )
assert( clamp_s16(000010) == 000010 )
assert( clamp_s16(000100) == 000100 )
assert( clamp_s16(001000) == 001000 )
assert( clamp_s16(010000) == 010000 )
assert( clamp_s16(100000) == 032766 )

assert( clamp_s16(-00000) == -00000 )
assert( clamp_s16(-00009) == -00009 )
assert( clamp_s16(-00099) == -00099 )
assert( clamp_s16(-00999) == -00999 )
assert( clamp_s16(-09999) == -09999 )
assert( clamp_s16(-99999) == -32767 )

minetest_find_nodes_in_area = minetest.find_nodes_in_area
minetest.find_nodes_in_area = function(minp, maxp, nodenames)
	local result1, result2
	if
		minp.x >= 32767 or minp.x <= -32768 or
		minp.y >= 32767 or minp.y <= -32768 or
		minp.z >= 32767 or minp.z <= -32768 or
		maxp.x >= 32767 or maxp.x <= -32768 or
		maxp.y >= 32767 or maxp.y <= -32768 or
		maxp.z >= 32767 or maxp.z <= -32768
	then
		minetest.log(
			"warning",
			"find_nodes_in_area() called with coords outside interval (-32768, 32767), clamping arguments: " ..
			"minp { x=" .. minp.x .. ", y=" .. minp.y .. " z=" .. minp.z .. " } " ..
			"maxp { x=" .. maxp.x .. ", y=" .. maxp.y .. " z=" .. maxp.z .. " } "
		)
		result1, result2 = minetest_find_nodes_in_area(
			{ x=clamp_s16(minp.x), y=clamp_s16(minp.y), z=clamp_s16(minp.z) },
			{ x=clamp_s16(maxp.x), y=clamp_s16(maxp.y), z=clamp_s16(maxp.z) },
			nodenames
		)
	else
		result1, result2 = minetest_find_nodes_in_area(
			minp,
			maxp,
			nodenames
		)
	end
end

minetest.register_chatcommand("areatest", {
func = function(name)
for x = 32767, 32700, -10 do
	for y = -32768, -32700, 10 do
		for z = 32760, 32770, 0.1 do
			local radius = 8
			local pos = { x=x, y=y, z=z }
			local areamin = vector.subtract(pos, radius)
			local areamax = vector.add(pos, radius)
			local npos, nnum = minetest.find_nodes_in_area(
			   areamin,
			   areamax,
			   "air"
			)
		end
	end
end
end
})

```

**Comment by appgurueu** [comment]
That can be shortened quite a bit:
```lua
minetest_find_nodes_in_area = minetest.find_nodes_in_area
function minetest.find_nodes_in_area(minp, maxp, nodenames)
        local clamped_minp = vector.apply(minp, clamp_s16)
	local clamped_maxp = vector.apply(maxp, clamp_s16)
	if not (vector.equals(clamped_minp, minp) and vector.equals(clamped_maxp, maxp)) then
		minetest.log(
			"warning",
			"find_nodes_in_area() called with coords outside interval (-32768, 32767), clamping arguments: " ..
			"minp { x=" .. minp.x .. ", y=" .. minp.y .. " z=" .. minp.z .. " } " ..
			"maxp { x=" .. maxp.x .. ", y=" .. maxp.y .. " z=" .. maxp.z .. " } "
		)
		
	end
		
	return minetest_find_nodes_in_area(
			{ x=clamp_s16(minp.x), y=clamp_s16(minp.y), z=clamp_s16(minp.z) },
			{ x=clamp_s16(maxp.x), y=clamp_s16(maxp.y), z=clamp_s16(maxp.z) },
			nodenames
		)
end
```

**Comment by erlehmann** [comment]
@appgurueu your solution has worse runtime behaviour though – it calls clamp_s16 6 times without any need.

In the overwhelming majority of cases the clamping is not necessary, i.e. you are wasting 6 function calls.

In my tests those unnecessarry function calls make the `/areatest` command take about 1.2s instead of 1.1s.

I *strongly* doubt that making `find_nodes_in_area()` take about 10% more time to have a bit less code is worth it.

**Comment by appgurueu** [comment]
> @appgurueu your solution has worse runtime behaviour though – it calls clamp_s16 6 times without any need.

Not really. It *might* differ by a small constant additive constant. In the worst case however, your solution performs more comparisons. But really, it doesn't matter - I assure you that all of this is in the < 1 microsecond range on reasonably performant devices.

**Comment by erlehmann** [comment]
@appgurueu can you find a difference in runtime between the following three commands? On my machine they all take about 13 to 15 seconds to run:
```
function clamp(value, min, max)
	assert( min < max )
	if value < min then
		return min
	end
	if value > max then
		return max
	end
	return value
end

assert( clamp(000, -100, 100) == 000 )
assert( clamp(999, -100, 100) == 100 )
assert( clamp(999, -999, 999) == 999 )
assert( clamp(998, 999, 1999) == 999 )
assert( clamp(999, 999, 1999) == 999 )

function clamp_s16(value)
	-- seems minetest hangs on -32768 and 32767
	return clamp(value, -32767, 32766)
end

assert( clamp_s16(000000) == 000000 )
assert( clamp_s16(000001) == 000001 )
assert( clamp_s16(000010) == 000010 )
assert( clamp_s16(000100) == 000100 )
assert( clamp_s16(001000) == 001000 )
assert( clamp_s16(010000) == 010000 )
assert( clamp_s16(100000) == 032766 )

assert( clamp_s16(-00000) == -00000 )
assert( clamp_s16(-00009) == -00009 )
assert( clamp_s16(-00099) == -00099 )
assert( clamp_s16(-00999) == -00999 )
assert( clamp_s16(-09999) == -09999 )
assert( clamp_s16(-99999) == -32767 )

minetest_find_nodes_in_area = minetest.find_nodes_in_area
minetest.find_nodes_in_area1 = function(minp, maxp, ...)
	local result1, result2
	if
		minp.x >= 32767 or minp.x <= -32768 or
		minp.y >= 32767 or minp.y <= -32768 or
		minp.z >= 32767 or minp.z <= -32768 or
		maxp.x >= 32767 or maxp.x <= -32768 or
		maxp.y >= 32767 or maxp.y <= -32768 or
		maxp.z >= 32767 or maxp.z <= -32768
	then
		minetest.log(
			"warning",
			"find_nodes_in_area() called with coords outside interval (-32768, 32767), clamping arguments: " ..
			"minp { x=" .. minp.x .. ", y=" .. minp.y .. " z=" .. minp.z .. " } " ..
			"maxp { x=" .. maxp.x .. ", y=" .. maxp.y .. " z=" .. maxp.z .. " } "
		)
		result1, result2 = minetest_find_nodes_in_area(
			{ x=clamp_s16(minp.x), y=clamp_s16(minp.y), z=clamp_s16(minp.z) },
			{ x=clamp_s16(maxp.x), y=clamp_s16(maxp.y), z=clamp_s16(maxp.z) },
			...
		)
	else
		result1, result2 = minetest_find_nodes_in_area(
			minp,
			maxp,
			...
		)
	end
end

minetest.find_nodes_in_area2 = function(minp, maxp, ...)
	local result1, result2
	local clamped_minp = vector.apply(minp, clamp_s16)
	local clamped_maxp = vector.apply(maxp, clamp_s16)
	if not (vector.equals(clamped_minp, minp) and vector.equals(clamped_maxp, maxp)) then
		minetest.log(
			"warning",
			"find_nodes_in_area() called with coords outside interval (-32768, 32767), clamping arguments: " ..
			"minp { x=" .. minp.x .. ", y=" .. minp.y .. " z=" .. minp.z .. " } " ..
			"maxp { x=" .. maxp.x .. ", y=" .. maxp.y .. " z=" .. maxp.z .. " } "
		)
	end
	result1, result2 =  minetest_find_nodes_in_area(
		{ x=clamp_s16(minp.x), y=clamp_s16(minp.y), z=clamp_s16(minp.z) },
		{ x=clamp_s16(maxp.x), y=clamp_s16(maxp.y), z=clamp_s16(maxp.z) },
		...
	)
	return result1, result2
end

minetest.find_nodes_in_area3 = function(minp, maxp, ...)
	local result1, result2
	local clamped_minp = vector.apply(minp, clamp_s16)
	local clamped_maxp = vector.apply(maxp, clamp_s16)
	if not (vector.equals(clamped_minp, minp) and vector.equals(clamped_maxp, maxp)) then
		minetest.log(
			"warning",
			"find_nodes_in_area() called with coords outside interval (-32768, 32767), clamping arguments: " ..
			"minp { x=" .. minp.x .. ", y=" .. minp.y .. " z=" .. minp.z .. " } " ..
			"maxp { x=" .. maxp.x .. ", y=" .. maxp.y .. " z=" .. maxp.z .. " } "
		)
		result1, result2 = minetest_find_nodes_in_area(
			{ x=clamp_s16(minp.x), y=clamp_s16(minp.y), z=clamp_s16(minp.z) },
			{ x=clamp_s16(maxp.x), y=clamp_s16(maxp.y), z=clamp_s16(maxp.z) },
			...
		)
	else
		result1, result2 = minetest_find_nodes_in_area(
			minp,
			maxp,
			...
		)
	end
	return result1, result2
end

minetest.register_chatcommand("areatest1", {
func = function(name)
start = os.clock()
for x = 32767, 32700, -10 do
	for y = -32768, -32700, 10 do
		for z = 32760, 32770, 0.1 do
			local radius = 8
			local pos = { x=x, y=y, z=z }
			local areamin = vector.subtract(pos, radius)
			local areamax = vector.add(pos, radius)
			local npos, nnum = minetest.find_nodes_in_area1(
			   areamin,
			   areamax,
			   "air"
			)
		end
	end
end
fin = os.clock()
minetest.debug(fin - start)
end
})

minetest.register_chatcommand("areatest2", {
func = function(name)
start = os.clock()
for x = 32767, 32700, -10 do
	for y = -32768, -32700, 10 do
		for z = 32760, 32770, 0.1 do
			local radius = 8
			local pos = { x=x, y=y, z=z }
			local areamin = vector.subtract(pos, radius)
			local areamax = vector.add(pos, radius)
			local npos, nnum = minetest.find_nodes_in_area2(
			   areamin,
			   areamax,
			   "air"
			)
		end
	end
end
fin = os.clock()
minetest.debug(fin - start)
end
})

minetest.register_chatcommand("areatest3", {
func = function(name)
start = os.clock()
for x = 32767, 32700, -10 do
	for y = -32768, -32700, 10 do
		for z = 32760, 32770, 0.1 do
			local radius = 8
			local pos = { x=x, y=y, z=z }
			local areamin = vector.subtract(pos, radius)
			local areamax = vector.add(pos, radius)
			local npos, nnum = minetest.find_nodes_in_area3(
			   areamin,
			   areamax,
			   "air"
			)
		end
	end
end
fin = os.clock()
minetest.debug(fin - start)
end
})
```

**Comment by appgurueu** [comment]
Consider localizing some functions if you really want to improve performance.

**Comment by erlehmann** [comment]
> Consider localizing some functions if you really want to improve performance.

Please elaborate. I have no idea what it means.

**Comment by appgurueu** [comment]
> > Consider localizing some functions if you really want to improve performance.
> 
> Please elaborate. I have no idea what it means.

You are using only global variables. Apart from the obvious global pollution, this is detrimental to performance: Every usage of a global variable requires indexing the hash part of the global table `_G`. `clamp` is only syntactic sugar for `_G["clamp"]`.

**Comment by Festus1965** [comment]
crashed on 01.03.2022 also
- Aaia Thailand on 5.2.0
- and view minutes then
- - Your-Land.de
- as Gamer Johnson12 from USA, California

**Comment by SmallJoker** [comment]
@Festus1965 The fix for this issue was merged on 26 Nov 2021, which is 5.5.0-dev. Please try to reproduce this issue on 5.5.0 stable.
